core/src/main/java/org/apache/iceberg/io/ResolvingFileIO.java

```diff
@@ -174,7 +176,7 @@ DelegateFileIO io(String location) {
     return ioInstances.computeIfAbsent(
         impl,
         key -> {
-          Configuration conf = hadoopConf.get();
+          Configuration conf = Optional.ofNullable(hadoopConf).map(Supplier::get).orElse(null);
```

Comment:
given that `ResolvingFileIO` is implementing `Configurable` from Hadoop it actually requires the conf to be set via `setConf` but I agree that this is a bit problematic for FileIO instances that don't need this, like `S3FileIO`.

Reply:
I do feel like it would be better to add this null handling to `getConf()` directly and then just call `getConf()` here